factor/totp/classes/factor.php

```diff
@@ -184,19 +184,31 @@ public function login_form_definition($mform) {
     public function login_form_validation($data) {
         $factors = $this->get_active_user_factors();
         $result = array('verificationcode' => get_string('error:wrongverification', 'factor_totp'));
+        $windowconfig = get_config('factor_totp', 'window');
 
         foreach ($factors as $factor) {
             $totp = TOTP::create($factor->secret);
-
+            // Convert seconds to windows.
+            $window = (int) floor($windowconfig / $totp->getPeriod());
             // First check if this code matches the last verified timestamp.
             $lastverified = $this->get_lastverified($factor->id);
-            if ($totp->verify($data['verificationcode'], $lastverified, 1)) {
+            if ($totp->verify($data['verificationcode'], $lastverified, $window)) {
                 return array('verificationcode' => get_string('error:codealreadyused', 'factor_totp'));
             }
 
-            if ($totp->verify($data['verificationcode'], time(), 1)) {
-                $result = array();
+            $pasttimestamp = time() - (5 * $totp->getPeriod());
```

Comment:
you've got 5 as a magic number in 4 places, can you just pull that out into a shared var with a good name